easypost/resource.py

```diff
@@ -0,0 +1,284 @@
+
+    @classmethod
+    def class_url(cls):
+        cls_name = cls.snakecase_name()
+        if cls_name[-1:] in ("s", "h"):
+            return "/%ses" % cls_name
+        else:
+            return "/%ss" % cls_name
+
+    def instance_url(self):
+        easypost_id = self.get("id")
+        if not easypost_id:
+            raise Error("%s instance has invalid ID: %r" % (type(self).__name__, easypost_id))
+        return "%s/%s" % (self.class_url(), easypost_id)
+
+
+class AllResource(Resource):
+    @classmethod
+    def all(cls, api_key=None, **params):
+        requestor = Requestor(api_key)
+        url = cls.class_url()
+        response, api_key = requestor.request("get", url, params)
+        return util.convert_to_easypost_object(response, api_key)
+
+
+class CreateResource(Resource):
+    @classmethod
+    def create(cls, api_key=None, **params):
+        requestor = Requestor(api_key)
+        url = cls.class_url()
+        wrapped_params = {cls.snakecase_name(): params}
+        response, api_key = requestor.request("post", url, wrapped_params)
+        return util.convert_to_easypost_object(response, api_key)
+
+
+class UpdateResource(Resource):
+    def save(self):
+        if self._unsaved_values:
+            requestor = Requestor(self._api_key)
+            params = {}
+            for k in self._unsaved_values:
+                params[k] = getattr(self, k)
+                if type(params[k]) is EasyPostObject:
+                    params[k] = params[k].flatten_unsaved()
+            params = {self.snakecase_name(): params}
+            url = self.instance_url()
+            response, api_key = requestor.request("put", url, params)
+            self.refresh_from(response, api_key)
+
+        return self
+
+
+class DeleteResource(Resource):
+    def delete(self, **params):
+        requestor = Requestor(self._api_key)
+        url = self.instance_url()
+        response, api_key = requestor.request("delete", url, params)
+        self.refresh_from(response, api_key)
+        return self
+
+
+class Requestor(object):
+    def __init__(self, local_api_key=None):
+        self._api_key = local_api_key
+
+    @classmethod
+    def _objects_to_ids(cls, param):
+        if isinstance(param, Resource):
+            return {"id": param.id}
+        elif isinstance(param, dict):
+            return {k: cls._objects_to_ids(v) for k, v in param.items()}
+        elif isinstance(param, list):
+            return [cls._objects_to_ids(v) for v in param]
+        else:
+            return param
+
+    def request(self, method, url, params=None, api_key_required=True):
+        if params is None:
+            params = {}
+        http_body, http_status, my_api_key = self.request_raw(method, url, params, api_key_required)
+        response = self.interpret_response(http_body, http_status)
+        return response, my_api_key
+
+    def request_raw(self, method, url, params=None, api_key_required=True):
+        from easypost import (
```

Comment:
We should import everything at the top of a file when possible. Importing elsewhere (unless for a specific reason such as wrapping in a try/catch like we do for the `google` package is typically not the best option because it hides away the imports inside the logic.

Reply:
I tried that, this would cause `partially initialized module 'easypost'` which most likely is circular import, in the `__init__.py` we have modules that are importing resource.py.

Reply:
Ugh. For now, please add an comment above this import statement detailing why it's here. Something like `importing here to avoid circular imports` is fine.